From bafd1ecf8b2d29cb7b2f057d66925105ebce7793 Mon Sep 17 00:00:00 2001
From: jim <jim.yuan@supermicro.com>
Date: Fri, 28 Apr 2017 13:24:48 -0700
Subject: [PATCH 1/3] change esel command from AMI to IBM 0x3a.

---
 platforms/astbmc/p9dsu.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/platforms/astbmc/p9dsu.c b/platforms/astbmc/p9dsu.c
index dbfd802..834a830 100644
--- a/platforms/astbmc/p9dsu.c
+++ b/platforms/astbmc/p9dsu.c
@@ -38,13 +38,18 @@ static bool p9dsu_probe(void)
 	return true;
 }
 
+static const struct bmc_platform astbmc_smc = {
+	.name = "SMC",
+	.ipmi_oem_partial_add_esel   = IPMI_CODE(0x3a, 0xf0),
+};
+
 DECLARE_PLATFORM(p9dsu) = {
 	.name			= "p9dsu",
 	.probe			= p9dsu_probe,
 	.init			= astbmc_init,
 	.start_preload_resource	= flash_start_preload_resource,
 	.resource_loaded	= flash_resource_loaded,
-	.bmc			= NULL, /* FIXME: Add openBMC */
+	.bmc			= &astbmc_smc,  /* FIXME: Add openBMC */
 	.pci_get_slot_info	= slot_table_get_slot_info,
 	.pci_probe_complete	= check_all_slot_table,
 	.cec_power_down         = astbmc_ipmi_power_down,
-- 
2.9.2.windows.1

