From b773ff67fc933d340ee4e462a17e139072ec8fa7 Mon Sep 17 00:00:00 2001
From: jim <jim.yuan@supermicro.com>
Date: Wed, 18 Oct 2017 16:05:24 -0700
Subject: [PATCH] Fix hdatBmcFamily string is overwritten. Issue 366.

---
 src/usr/hdat/hdatspsubsys.C | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/usr/hdat/hdatspsubsys.C b/src/usr/hdat/hdatspsubsys.C
index 172bc99..1119e77 100755
--- a/src/usr/hdat/hdatspsubsys.C
+++ b/src/usr/hdat/hdatspsubsys.C
@@ -421,9 +421,9 @@ errlHndl_t HdatSpSubsys::hdatFillDataPtrs()
 
            strcpy( iv_impl.hdatBmcFamily , l_bmcManufacturer);
            strcat( iv_impl.hdatBmcFamily , ",");
-           strcpy( iv_impl.hdatBmcFamily , l_bmcHwChip);
+           strcat( iv_impl.hdatBmcFamily , l_bmcHwChip);
            strcat( iv_impl.hdatBmcFamily , ",");
-           strcpy( iv_impl.hdatBmcFamily , l_bmcSw);
+           strcat( iv_impl.hdatBmcFamily , l_bmcSw);
         }
         else
         {
-- 
2.9.2.windows.1

